In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt

In [2]:
data = 'https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv'

In [3]:
!wget $data -O data-homework.csv

--2023-10-03 02:55:48--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1475504 (1.4M) [text/plain]
Saving to: ‘data-homework.csv’

data-homework.csv   100%[===================>]   1.41M  --.-KB/s    in 0.06s   

2023-10-03 02:55:48 (22.9 MB/s) - ‘data-homework.csv’ saved [1475504/1475504]



In [4]:
df = pd.read_csv('data-homework.csv')
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [5]:
df.columns

Index(['Make', 'Model', 'Year', 'Engine Fuel Type', 'Engine HP',
       'Engine Cylinders', 'Transmission Type', 'Driven_Wheels',
       'Number of Doors', 'Market Category', 'Vehicle Size', 'Vehicle Style',
       'highway MPG', 'city mpg', 'Popularity', 'MSRP'],
      dtype='object')

In [6]:
columns =['Make', 'Model', 'Year','Engine HP',
       'Engine Cylinders', 'Transmission Type',
        'Vehicle Style','highway MPG', 'city mpg','MSRP']

In [7]:
df_sample = df[columns]
df_sample.head()

Make       Model  Year  Engine HP  Engine Cylinders Transmission Type  \
0  BMW  1 Series M  2011      335.0               6.0            MANUAL   
1  BMW    1 Series  2011      300.0               6.0            MANUAL   
2  BMW    1 Series  2011      300.0               6.0            MANUAL   
3  BMW    1 Series  2011      230.0               6.0            MANUAL   
4  BMW    1 Series  2011      230.0               6.0            MANUAL   

  Vehicle Style  highway MPG  city mpg   MSRP  
0         Coupe           26        19  46135  
1   Convertible           28        19  40650  
2         Coupe           28        20  36350  
3         Coupe           28        18  29450  
4   Convertible           28        18  34500

In [8]:
df_sample.columns = df_sample.columns.str.lower().str.replace(' ', '_')

In [9]:
df_sample = df_sample.fillna(0)

In [10]:
df_sample.dtypes

make                  object
model                 object
year                   int64
engine_hp            float64
engine_cylinders     float64
transmission_type     object
vehicle_style         object
highway_mpg            int64
city_mpg               int64
msrp                   int64
dtype: object

In [11]:
df_sample.isnull().sum()

make                 0
model                0
year                 0
engine_hp            0
engine_cylinders     0
transmission_type    0
vehicle_style        0
highway_mpg          0
city_mpg             0
msrp                 0
dtype: int64

In [12]:
df_sample.rename(columns ={'msrp':'price'},inplace =True)

In [13]:
df_sample.head()

make       model  year  engine_hp  engine_cylinders transmission_type  \
0  BMW  1 Series M  2011      335.0               6.0            MANUAL   
1  BMW    1 Series  2011      300.0               6.0            MANUAL   
2  BMW    1 Series  2011      300.0               6.0            MANUAL   
3  BMW    1 Series  2011      230.0               6.0            MANUAL   
4  BMW    1 Series  2011      230.0               6.0            MANUAL   

  vehicle_style  highway_mpg  city_mpg  price  
0         Coupe           26        19  46135  
1   Convertible           28        19  40650  
2         Coupe           28        20  36350  
3         Coupe           28        18  29450  
4   Convertible           28        18  34500

In [14]:
most_frequent = df_sample['transmission_type'].mode()
most_frequent.values[0]

'AUTOMATIC'

In [15]:
df_sample['engine_hp'].corr(df_sample.year)

0.33871418476244686

In [16]:
df_sample['engine_hp'].corr(df_sample.engine_cylinders)

0.7748509807813194

In [17]:
df_sample['highway_mpg'].corr(df_sample.engine_cylinders)

-0.6145414173953334

In [18]:
df_sample['highway_mpg'].corr(df_sample.city_mpg)

0.886829496259137

For correlation between engine_hp and year is 0.33871418476244686.
For correlation between engine_hp and engine_cylinders is 0.7748509807813194.
For correlation between highway_mpg and engine_cylinders is -0.6145414173953334.
For correlation between highway_mpg and city_mpg is 0.886829496259137.
The highest correlation is highway_mpg and city_mpg

In [19]:
df_sample.price.mean()

40594.737032063116

In [20]:
df_sample['above_average'] = (df_sample.price > df_sample.price.mean()).astype(int)
df_sample.above_average

0        1
1        1
2        0
3        0
4        0
        ..
11909    1
11910    1
11911    1
11912    1
11913    0
Name: above_average, Length: 11914, dtype: int64

In [21]:
df_sample.head()

make       model  year  engine_hp  engine_cylinders transmission_type  \
0  BMW  1 Series M  2011      335.0               6.0            MANUAL   
1  BMW    1 Series  2011      300.0               6.0            MANUAL   
2  BMW    1 Series  2011      300.0               6.0            MANUAL   
3  BMW    1 Series  2011      230.0               6.0            MANUAL   
4  BMW    1 Series  2011      230.0               6.0            MANUAL   

  vehicle_style  highway_mpg  city_mpg  price  above_average  
0         Coupe           26        19  46135              1  
1   Convertible           28        19  40650              1  
2         Coupe           28        20  36350              0  
3         Coupe           28        18  29450              0  
4   Convertible           28        18  34500              0

# Spliting the datasets

In [22]:
from sklearn.model_selection import train_test_split

In [23]:
df_full_train, df_test = train_test_split(df_sample, test_size= 0.2, random_state= 42)

In [24]:
len(df_full_train), len(df_test)

(9531, 2383)

In [25]:
df_train, df_val = train_test_split(df_full_train, test_size= 0.25, random_state= 42)

In [26]:
len(df_train),len(df_test),len(df_val)

(7148, 2383, 2383)

In [27]:
df_train.reset_index(drop = True)
df_val.reset_index(drop =True)
df_test.reset_index(drop = True)

make             model  year  engine_hp  engine_cylinders  \
0            GMC          Envoy XL  2005      275.0               6.0   
1     Volkswagen            Passat  2016      170.0               4.0   
2          Honda           Odyssey  2016      248.0               6.0   
3      Chevrolet             Cruze  2015      138.0               4.0   
4          Volvo               740  1991      162.0               4.0   
...          ...               ...   ...        ...               ...   
2378      Nissan              350Z  2007      306.0               6.0   
2379        Ford        Expedition  2016      365.0               6.0   
2380       Acura               ILX  2015      150.0               4.0   
2381  Volkswagen  Jetta SportWagen  2014      170.0               5.0   
2382         BMW                X5  2015      255.0               6.0   

     transmission_type      vehicle_style  highway_mpg  city_mpg  price  \
0            AUTOMATIC            4dr SUV           18        13  29695   
1            AUTOMATIC              Sedan           38        25  30495   
2            AUTOMATIC  Passenger Minivan           28        19  37650   
3               MANUAL              Sedan           36        25  16170   
4            AUTOMATIC              Sedan           20        17   2000   
...                ...                ...          ...       ...    ...   
2378            MANUAL        Convertible           24        17  35550   
2379         AUTOMATIC            4dr SUV           20        15  48360   
2380         AUTOMATIC              Sedan           35        24  31750   
2381            MANUAL              Wagon           33        23  20995   
2382         AUTOMATIC            4dr SUV           31        24  57700   

      above_average  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  
...             ...  
2378              0  
2379              1  
2380              0  
2381              0  
2382              1  

[2383 rows x 11 columns]

In [28]:
y_train = df_train.above_average.values
y_val = df_val.above_average.values
y_test = df_test.above_average.values

In [29]:
del df_train['above_average']
del df_val['above_average']
del df_test['above_average']

# Mutual information score

In [30]:
from sklearn.metrics import mutual_info_score

In [31]:
df_full_train.above_average

3181     1
5357     1
4874     0
8102     0
10400    0
        ..
11284    0
5191     0
5390     0
860      1
7270     0
Name: above_average, Length: 9531, dtype: int64

In [32]:
def mutual_info_churn_score(series):
    return mutual_info_score(series, df_full_train.above_average)

In [33]:
column_2 =['make','model','transmission_type','vehicle_style']

In [34]:
mi = df_full_train[column_2].apply(mutual_info_churn_score)
round(mi,2)

make                 0.24
model                0.46
transmission_type    0.02
vehicle_style        0.08
dtype: float64

# One hot encoding

In [35]:
from sklearn.feature_extraction import DictVectorizer

In [36]:
df_train.dtypes

make                  object
model                 object
year                   int64
engine_hp            float64
engine_cylinders     float64
transmission_type     object
vehicle_style         object
highway_mpg            int64
city_mpg               int64
price                  int64
dtype: object

In [82]:
categorical = ['make','model','transmission_type','vehicle_style']

In [83]:
numerical = ['year','engine_hp','engine_cylinders','highway_mpg','city_mpg']

In [84]:
train_dicts = df_train[categorical + numerical].to_dict(orient = 'records')

In [85]:
dv = DictVectorizer(sparse = False)

In [86]:
X_train = dv.fit_transform(train_dicts)

In [87]:
val_dicts = df_val[categorical + numerical].to_dict(orient = 'records')

In [88]:
X_val = dv.transform(val_dicts)

In [89]:
from sklearn.linear_model import LogisticRegression

In [90]:
model = LogisticRegression(solver='liblinear', C=10, max_iter=1000, random_state=42)
model.fit(X_train,y_train)

LogisticRegression(C=10, max_iter=1000, random_state=42, solver='liblinear')

In [91]:
y_pred = model.predict_proba(X_val)[:,1]
y_pred

array([4.17000993e-04, 9.97836444e-01, 6.44601930e-05, ...,
       1.57073539e-04, 9.93752243e-01, 9.90597855e-01])

In [92]:
price_decision = (y_pred >= 0.5)
price_decision

array([False,  True, False, ..., False,  True,  True])

In [93]:
df_predict = pd.DataFrame()
df_predict['probability'] = y_pred
df_predict['prediction'] = price_decision.astype(int)
df_predict['actual'] = y_val

In [94]:
df_predict['correct'] = df_predict.prediction == df_predict.actual

In [95]:
original_accuracy = df_predict['correct'].mean()
original_accuracy

0.946286193873269

## Train the dataset with removing year

In [96]:
year = ['make','model','transmission_type','vehicle_style','engine_hp','engine_cylinders','highway_mpg','city_mpg']

In [97]:
train_year_dicts = df_train[year].to_dict(orient = 'records')

In [98]:
dv = DictVectorizer(sparse = False)

In [99]:
X_year_train = dv.fit_transform(train_year_dicts)

In [100]:
val_year_dicts = df_val[year].to_dict(orient = 'records')

In [101]:
X_year_val = dv.transform(val_year_dicts)

In [102]:
model = LogisticRegression(solver='liblinear', C=10, max_iter=1000, random_state=42)
model.fit(X_year_train,y_train)

LogisticRegression(C=10, max_iter=1000, random_state=42, solver='liblinear')

In [104]:
y_year_pred = model.predict_proba(X_year_val)[:,1]
y_year_pred

array([3.61861150e-04, 9.97007937e-01, 5.81509772e-05, ...,
       3.09914294e-04, 9.95857396e-01, 9.93490329e-01])

In [105]:
price_year_decision = (y_year_pred >= 0.5)
price_year_decision

array([False,  True, False, ..., False,  True,  True])

In [107]:
df_year_pred = pd.DataFrame()
df_year_pred['probability'] = y_year_pred
df_year_pred['prediction'] = price_year_decision.astype(int)
df_year_pred['actual'] = y_val

In [108]:
df_year_pred['correct'] = df_year_pred.prediction == df_year_pred.actual

In [110]:
without_year_acc = df_year_pred['correct'].mean()

In [111]:
diff_acc_year_and_actual = original_accuracy - without_year_acc

In [112]:
diff_acc_year_and_actual

-0.0012589173310952884

# Train the dataset without engine_hp

In [113]:
engine = ['make','model','transmission_type','vehicle_style','year','engine_cylinders','highway_mpg','city_mpg']

In [114]:
train_engine_dicts = df_train[engine].to_dict(orient = 'records')
dv = DictVectorizer(sparse = False)
X_engine_train = dv.fit_transform(train_engine_dicts)
val_engine_dicts = df_val[engine].to_dict(orient = 'records')
X_engine_val = dv.transform(val_engine_dicts)

In [115]:
model = LogisticRegression(solver='liblinear', C=10, max_iter=1000, random_state=42)
model.fit(X_engine_train,y_train)

LogisticRegression(C=10, max_iter=1000, random_state=42, solver='liblinear')

In [116]:
y_engine_pred = model.predict_proba(X_engine_val)[:,1]
y_engine_pred

array([7.30759367e-04, 9.13342029e-01, 1.13812346e-03, ...,
       1.43472636e-02, 9.88938084e-01, 9.71354289e-01])

In [117]:
price_engine_decision = (y_engine_pred >= 0.5)
price_engine_decision

array([False,  True, False, ..., False,  True,  True])

In [118]:
df_engine_pred = pd.DataFrame()
df_engine_pred['probability'] = y_engine_pred
df_engine_pred['prediction'] = price_engine_decision.astype(int)
df_engine_pred['actual'] = y_val
df_engine_pred['correct'] = df_engine_pred.prediction == df_engine_pred.actual

In [120]:
without_engine = df_engine_pred['correct'].mean()
diff_accu_engine_and_actual = original_accuracy - without_engine
diff_accu_engine_and_actual

0.022240872849349502

# Train the dataset without transmission types

In [121]:
trans = ['make','model','vehicle_style','year','engine_hp','engine_cylinders','highway_mpg','city_mpg']

In [122]:
train_trans_dicts = df_train[trans].to_dict(orient = 'records')
dv = DictVectorizer(sparse = False)
X_trans_train = dv.fit_transform(train_trans_dicts)
val_trans_dicts = df_val[trans].to_dict(orient = 'records')
X_trans_val = dv.transform(val_trans_dicts)

In [125]:
model = LogisticRegression(solver='liblinear', C=10, max_iter=1000, random_state=42)
model.fit(X_trans_train,y_train)

LogisticRegression(C=10, max_iter=1000, random_state=42, solver='liblinear')

In [126]:
y_trans_pred = model.predict_proba(X_trans_val)[:,1]
price_trans_decision = (y_trans_pred >= 0.5)

In [127]:
df_trans_pred = pd.DataFrame()
df_trans_pred['probability'] = y_trans_pred
df_trans_pred['prediction'] = price_trans_decision.astype(int)
df_trans_pred['actual'] = y_val
df_trans_pred['correct'] = df_trans_pred.prediction == df_engine_pred.actual

In [128]:
without_trans = df_trans_pred['correct'].mean()
diff_accu_trans_and_actual = original_accuracy - without_trans
diff_accu_trans_and_actual

0.0012589173310951773

# Train the dataset without city_mpg

In [133]:
city = ['make','model','vehicle_style','year','engine_hp','engine_cylinders','transmission_type','highway_mpg']

In [134]:
train_city_dicts = df_train[city].to_dict(orient = 'records')
dv = DictVectorizer(sparse = False)
X_city_train = dv.fit_transform(train_city_dicts)
val_city_dicts = df_val[city].to_dict(orient = 'records')
X_city_val = dv.transform(val_city_dicts)

In [137]:
model = LogisticRegression(solver='liblinear', C=10, max_iter=1000, random_state=42)
model.fit(X_city_train,y_train)

LogisticRegression(C=10, max_iter=1000, random_state=42, solver='liblinear')

In [138]:
y_city_pred = model.predict_proba(X_city_val)[:,1]
price_city_decision = (y_city_pred >= 0.5)

In [139]:
df_city_pred = pd.DataFrame()
df_city_pred['probability'] = y_city_pred
df_city_pred['prediction'] = price_city_decision.astype(int)
df_city_pred['actual'] = y_val
df_city_pred['correct'] = df_city_pred.prediction == df_city_pred.actual

In [140]:
without_city = df_city_pred['correct'].mean()
diff_accu_city_and_actual = original_accuracy - without_city
diff_accu_city_and_actual

0.0

# Train the dataset with the target price




In [153]:
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error

In [144]:
y_price_train = np.log1p(df_train.price.values)
y_price_val = np.log1p(df_val.price.values)
y_price_test = np.log1p(df_test.price.values)


In [145]:
del df_train['price']
del df_val['price']
del df_test['price']

In [147]:
final_train_dicts = df_train.to_dict(orient = 'records')
dv = DictVectorizer()
X_final_train = dv.fit_transform(final_train_dicts)
val_final_dicts = df_val.to_dict(orient = 'records')
X_final_val = dv.transform(val_final_dicts)


In [149]:
alphas = [0,0.01,0.1,1,10]

In [155]:
for alpha in alphas:
    model = Ridge(alpha = alpha, solver ='sag', random_state=42)
    model.fit(X_final_train, y_price_train)
    y_final_pred = model.predict(X_final_val)
    rmse = round(np.sqrt(mean_squared_error(y_price_val, y_final_pred)),2)
    print('For alpha',alpha,' : Root mean squared error value is: ',rmse)

For alpha 0  : Root mean squared error value is:  0.25
For alpha 0.01  : Root mean squared error value is:  0.25
For alpha 0.1  : Root mean squared error value is:  0.26
For alpha 1  : Root mean squared error value is:  0.26
For alpha 10  : Root mean squared error value is:  0.34
